### PR TITLE
[aptos-vm] implement new gas metering system

### PR DESCRIPTION
**Pull request by vgao1996**
This implements a proper gas metering system for the Aptos blockchain. More specifically, this
1. Defines a new extensible on-chain gas schedule
2. Introduces an `aptos-gas` crate 
  a. Implements a new gas meter and defines the formulae the instruction costs 
  b. Defines a bi-directional mapping between the on-chain gas schedule and the rust representation
  c. Defines an initial gas schedule (the values are currently just placeholders)
3. Made `framework` natives customizable in terms of gas. 
4. Moved `code::from_bytes` into a new module `util` so it can be shared within the aptos framework
5. Moved the native functions `transaction_context::get_script_hash` and `code::request_publish` into `framework` for better organization
  a. This PR also fixes `code::request_publish` so that it properly returns `NativeResult::err(...)` when given invalid user input, instead of raising an internal error.

TODOs:
1. Documentation
2. ~~Code formatting~~
3. ~~Tests~~
4. ~~Lints~~
5. Other cleanups
6. ~~APIs of the `gas_schedule` move module~~
7. Storage-related gas parameters

---
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/2358)

### CODE REVIEW

File: aptos-move/framework/aptos-framework/sources/configs/gas_schedule.move [hunk 1]
```diff
@@ -1,88 +1,137 @@
...
+
+        assert!(exists<GasSchedule>(@aptos_framework), error::not_found(ECONFIG));
+
+        let gas_schedule = borrow_global_mut<GasSchedule>(@aptos_framework);
+        *gas_schedule = from_bytes(gas_schedule_blob);
+
+        reconfiguration::reconfigure();
+    }
+
+    /*
```
Comment (davidiw) [inline aptos-move/framework/aptos-framework/sources/configs/gas_schedule.move:124]: please delete legacy code.

File: aptos-move/framework/aptos-framework/sources/configs/gas_schedule.move [hunk 2]
```diff
@@ -1,88 +1,137 @@
...
+        else {
+            let idx = option::destroy_some(idx_opt);
+            let entry = vector::borrow_mut(&mut gas_schedule.entries, idx);
+            let old = entry.val;
+            entry.val = val;
+            option::some(old)
+        }
+    }
+
+    public fun remove(gas_schedule: &mut GasSchedule, key: &String): Option<u64> {
```
Comment (davidiw) [inline aptos-move/framework/aptos-framework/sources/configs/gas_schedule.move:80]: something doesn't quite feel right about these being exposed as public. I also didn't see them being used. Why do we want to expose much of any of this to the rest of the chain? I think we discussed trying to keep the data here readable for indexers, but not sure it needs to be useful for on-chain purposes outside of doing a bulk update via set_gas_schedule

File: aptos-move/framework/src/natives/account.rs [hunk 3]
```diff
@@ -3,28 +3,38 @@
...
+ * native fun create_address
+ *
+ *   gas cost: base_cost
+ *
+ **************************************************************************************************/
+#[derive(Debug, Clone)]
+pub struct CreateAddressGasParameters {
+    pub base_cost: u64,
+}
+
```
Comment (davidiw) [inline aptos-move/framework/src/natives/account.rs:26]: It seems like these are pretty constant typed... is there a way to unify these into a GasParameters struct or something that distinguishes the handful of types?

**Comment by wrwg** [comment]
Any reason why this is still draft? Some folks may not look until you remove the `draft` status.

File: aptos-move/framework/src/natives/code.rs [hunk 4]
```diff
@@ -124,24 +110,36 @@ pub struct PublishRequest {
...
+ * )
+ *
+ *   gas cost: base_cost + unit_cost * bytes_len
+ *
+ **************************************************************************************************/
+#[derive(Clone, Debug)]
+pub struct RequestPublishGasParameters {
+    pub base_cost: u64,
+    pub unit_cost: u64,
 }
```
Comment (davidiw) [inline aptos-move/framework/src/natives/code.rs:139]: now we have a type that expects base and unit... it seems like we could have this as part of the native registration process and make it simpler...

fortunately I think this is something that can be iterated on if it becomes to cumbersome to wrap up in the gas only effort.

File: aptos-move/aptos-gas/src/aptos_framework.rs [hunk 5]
```diff
@@ -0,0 +1,36 @@
+// Copyright (c) Aptos
+// SPDX-License-Identifier: Apache-2.0
+
+use framework::natives::GasParameters;
+
+crate::natives::define_gas_parameters_for_natives!(GasParameters, "aptos_framework", [
```
Comment (wrwg) [inline aptos-move/aptos-gas/src/aptos_framework.rs:6]: Loving this.

File: testsuite/smoke-test/Cargo.toml [hunk 6]
```diff
@@ -38,6 +38,8 @@ cached-framework-packages =  { path = "../../aptos-move/framework/cached-package
 consensus = { path = "../../consensus" }
 forge = { path = "../forge" }
 move-deps = { path = "../../aptos-move/move-deps" }
+aptos-gas = { path = "../../aptos-move/aptos-gas" }
+framework = { path = "../../aptos-move/framework" }
```
Comment (davidiw) [inline testsuite/smoke-test/Cargo.toml:42]: shame we lost our sorting of crates...

**Review by davidiw — COMMENTED** [review]
fantastic work! Only major gripe is I think we should make sure the gas schedule move file is simple and we can always expand upon it later. would be great to have just two types of gas parameters for natives as that's also something we can expand upon if we need it.

File: aptos-move/framework/src/natives/account.rs [hunk 7]
```diff
@@ -37,15 +47,65 @@ pub fn native_create_address(
...
+pub fn make_native_create_signer(gas_params: CreateSignerGasParameters) -> NativeFunction {
+    Arc::new(move |context, ty_args, args| {
+        native_create_signer(&gas_params, context, ty_args, args)
+    })
+}
+
+/***************************************************************************************************
+ * module
+ **************************************************************************************************/
```
Comment (wrwg) [inline aptos-move/framework/src/natives/account.rs:91]: Style: I would expect a newline here.

**Review by wrwg — APPROVED** [review]
I'm suggesting that we land this now but then follow up on individual gas formulas with the folks responsible for the particular design area. The general design was discussed and approved by the Move maintainers already.

**Comment by movekevin** [comment]
> I'm suggesting that we land this now but then follow up on individual gas formulas with the folks responsible for the particular design area. The general design was discussed and approved by the Move maintainers already.

Can you give me a day to review as well? Will look through by end of Monday at the latest.

File: aptos-move/framework/aptos-framework/sources/configs/gas_schedule.move (hunk 2 above)
Comment (wrwg) [reply aptos-move/framework/aptos-framework/sources/configs/gas_schedule.move:80]: I believe this needs to be public so we can define governance approved scripts which update the gas schedule. However, I think the caller needs to present a governance signer here and the code should abort if not.

File: aptos-move/framework/src/natives/account.rs (hunk 3 above)
Comment (wrwg) [reply aptos-move/framework/src/natives/account.rs:26]: There are pros and cons around such unification. (This was also discussed on the PR in move-language.) I'm tending to some symmetric boilerplate and dupllocation here. This might eventually be easier to maintain and extend instead of trying to optimize this.

**Comment by vgao1996** [comment]
@movekevin please take your time! It'll take me some time to deal with test failures and resolve conflicts. @wrwg this is the main reason I'm still marking this as a draft. You can find a full list of TODOs in the summary.

File: aptos-move/framework/aptos-framework/sources/configs/gas_schedule.move (hunk 2 above)
Comment (vgao1996) [reply aptos-move/framework/aptos-framework/sources/configs/gas_schedule.move:80]: Yeah @zekun000 had similar concerns and I agree we should take another look at the APIs this module should offer. To be clear, when I added these, I just simply thought they would be useful for creating governance scripts but didn't have the chance to carefully review and polish them yet.

How about we separate this into its own task:
1. Remove the functions that modify the on-chain gas schedule and land the PR
2. Discuss the design of these APIs and add them back later

File: aptos-move/aptos-vm/src/aptos_vm_impl.rs [hunk 8]
```diff
@@ -53,30 +48,45 @@ pub struct AptosVMImpl {
 impl AptosVMImpl {
     #[allow(clippy::new_without_default)]
     pub fn new<S: StateView>(state: &S) -> Self {
-        let inner = MoveVmExt::new()
+        let storage = RemoteStorage::new(state);
+
+        // TODO: this should not panic
```
[the remark this answers is not in the document]
Comment (vgao1996) [reply aptos-move/aptos-vm/src/aptos_vm_impl.rs:53]: The `aptos-vm` creation story is a bit messy. Ideally I also think it should either panic or return an `Err` if any of the inputs is bad. However if you take a look at the current code, you'll see the fields all have `Option` types and I haven't got the time to clean this up.

File: aptos-move/aptos-gas/src/instr.rs [hunk 9]
```diff
@@ -0,0 +1,330 @@
...
+    [branch, "branch", 1],
+    // stack
+    [pop, "pop", 1],
+    [ld_u8, "ld_u8", 1],
+    [ld_u64, "ld_u64", 1],
+    [ld_u128, "ld_u128", 1],
+    [ld_true, "ld_true", 1],
+    [ld_false, "ld_false", 1],
+    [ld_const_base, "ld_const.base", 1],
+    [ld_const_unit, "ld_const.unit", 1],
```
[the remark this answers is not in the document]
Comment (vgao1996) [reply aptos-move/aptos-gas/src/instr.rs:79]: I thought about this. I think it's possible but there can always be unexpected things with complex macros so I can't say for sure. I have some rough ideas I can try later.

Maybe another thing we could explore is procedural macro, but I feel like it may not work with upstream types (e.g. `move_stdlib::natives::GasParameters`)

File: aptos-move/framework/src/natives/account.rs (hunk 3 above)
Comment (vgao1996) [reply aptos-move/framework/src/natives/account.rs:26]: Yeah it's a choice between having less boilerplate code or more extensibility. If we feel this is a bit too much, we can always make fine adjustments in the future.

File: aptos-move/framework/src/natives/code.rs (hunk 4 above)
Comment (vgao1996) [reply aptos-move/framework/src/natives/code.rs:139]: I'd be cautious with requiring all native functions to have the `base_cost + unit_cost * len` model. It does work for the majority of them, but as long as there's one that doesn't, it will be a deal breaker. I'd also expect our gas formula to be more refined in the future, meaning that they will have more gas parameters.

File: aptos-move/aptos-vm/src/aptos_vm_impl.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (vgao1996) [reply aptos-move/aptos-vm/src/aptos_vm_impl.rs:52]: Okay, I think I now kinda understand why we have `Option`s in the first place: we expect the AptosVM to be able to execute a genesis writeset transaction. This code path does not require an on-chain gas schedule to be available.

TBH I'm not feeling very comfortable with this. It's conceptually hard to understand, and if you try to execute another transaction after the genesis, you'll end up with the VM panicking.

**Comment by vgao1996** [comment]
@movekevin let me know if you have any concerns. Otherwise I'll try to land it soon.

**Comment by movekevin** [comment]
> @movekevin let me know if you have any concerns. Otherwise I'll try to land it soon.

Looking at it. I can finish the review in ~2 hours (1 hour after this current SEV review)

File: aptos-move/aptos-gas/Cargo.toml [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+authors = ["Aptos Labs <opensource@aptoslabs.com>"]
+description = "Gas metering for the Aptos blockchain"
+repository = "https://github.com/aptos-labs/aptos-core"
+homepage = "https://aptoslabs.com"
+license = "Apache-2.0"
+publish = false
+edition = "2021"
+
+[dependencies]
+move-binary-format = { git = "https://github.com/move-language/move", rev = "7d28bfa484ef7d9ff225a748832de1cf1b67d77b" }
```
Comment (movekevin) [inline aptos-move/aptos-gas/Cargo.toml:13]: Why not use move-deps so the move dependency can be updated in the same place?

File: aptos-move/aptos-gas/src/aptos_framework.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (movekevin) [reply aptos-move/aptos-gas/src/aptos_framework.rs:6]: Is this file generated somehow? If not, I'm slightly concerned we might make a mistake (e.g. a typo) in the future when we add a new op type.

File: aptos-move/aptos-gas/src/gas_meter.rs [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+pub struct NativeGasParameters {
+    pub move_stdlib: move_stdlib::natives::GasParameters,
+    pub aptos_framework: framework::natives::GasParameters,
+}
+
+impl FromOnChainGasSchedule for NativeGasParameters {
+    fn from_on_chain_gas_schedule(gas_schedule: &BTreeMap<String, u64>) -> Option<Self> {
+        Some(Self {
+            move_stdlib: FromOnChainGasSchedule::from_on_chain_gas_schedule(gas_schedule)?,
+            aptos_framework: FromOnChainGasSchedule::from_on_chain_gas_schedule(gas_schedule)?,
```
Comment (movekevin) [inline aptos-move/aptos-gas/src/gas_meter.rs:38]: These look exactly the same to me. Why do they need to be 2 different variables?

File: aptos-move/aptos-gas/src/instr.rs [hunk 12]
```diff
@@ -0,0 +1,330 @@
...
+                )
+            }
+        })
+    }
+
+    pub(crate) fn instr_cost_with_size(&self, op: Opcodes, size: u64) -> PartialVMResult<u64> {
+        use Opcodes::*;
+
+        Ok(match op {
+            LD_CONST => self.ld_const_base + self.ld_const_unit * size,
```
Comment (movekevin) [inline aptos-move/aptos-gas/src/instr.rs:266]: All these entries look a bit redundant. Can we refactor this into a generic calculation function: base + unit * size?

File: aptos-move/aptos-vm/src/aptos_vm_impl.rs [hunk 13]
```diff
@@ -12,64 +12,82 @@ use crate::{
...
     chain_account_info: Option<ChainSpecificAccountInfo>,
 }
 
 impl AptosVMImpl {
     #[allow(clippy::new_without_default)]
     pub fn new<S: StateView>(state: &S) -> Self {
-        let inner = MoveVmExt::new()
+        let storage = RemoteStorage::new(state);
+
+        // TODO(Gas): this should not panic
```
Comment (movekevin) [inline aptos-move/aptos-vm/src/aptos_vm_impl.rs:53]: Are you going to address these todos in the follow up PRs?

File: aptos-move/framework/aptos-framework/sources/configs/gas_schedule.move [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+    use std::error;
+    use std::string::String;
+
+    use aptos_framework::reconfiguration;
+    use aptos_framework::system_addresses;
+    use aptos_framework::timestamp;
+    use aptos_framework::util::from_bytes;
+
+    /// Error with config
+    const ECONFIG: u64 = 0;
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/configs/gas_schedule.move:13]: nit: Can we not use 0? 0 should be reserved as it can be a default value.

File: aptos-move/framework/aptos-framework/sources/configs/gas_schedule.move [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+        key: String,
+        val: u64,
+    }
+
+    struct GasSchedule has key, copy, drop {
+        entries: vector<GasEntry>
+    }
+
+    public fun initialize(account: &signer, gas_schedule_blob: vector<u8>) {
+        timestamp::assert_genesis();
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/configs/gas_schedule.move:27]: Do we need this check? I'd ideally prefer not to have it unless it's absolutely critical. This doesn't look like it can be called more than once anyway (the assert on line 30 will fail)

File: aptos-move/framework/aptos-framework/sources/configs/gas_schedule.move [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+        system_addresses::assert_aptos_framework(account);
+
+        assert!(
+            !exists<GasSchedule>(@aptos_framework),
+            error::already_exists(ECONFIG)
+        );
+
+        // TODO(Gas): check if gas schedule is consistent
+
+        let gas_schedule: GasSchedule = from_bytes(gas_schedule_blob);
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/configs/gas_schedule.move:37]: nit: This can be inlined into line 39

File: aptos-move/framework/aptos-framework/sources/configs/gas_schedule.move [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+
+        move_to(account, gas_schedule);
+    }
+
+    public entry fun set_gas_schedule(account: &signer, gas_schedule_blob: vector<u8>) acquires GasSchedule {
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/configs/gas_schedule.move:42]: This doesn't need to be an entry function as it's only callable as part of the governance process from a script anyway. Scripts can call non-entry public functions

File: aptos-move/framework/aptos-framework/sources/configs/gas_schedule.move [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+        timestamp::assert_operating();
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/configs/gas_schedule.move:43]: nit: Doesn't seem necessary to me?

File: aptos-move/framework/aptos-framework/sources/configs/gas_schedule.move [hunk 19]
```diff
@@ -0,0 +1,55 @@
...
+        system_addresses::assert_core_resource(account);
+
+        assert!(exists<GasSchedule>(@aptos_framework), error::not_found(ECONFIG));
+
+        // TODO(Gas): check if gas schedule is consistent
+
+        let gas_schedule = borrow_global_mut<GasSchedule>(@aptos_framework);
+        *gas_schedule = from_bytes(gas_schedule_blob);
+
+        reconfiguration::reconfigure();
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/configs/gas_schedule.move:53]: Why is this necessary? Does the adapter also rely on the new epoch event to reload gas schedule? If so, it seems like overloading of that event. Can we use a separate, new event?

**Review by movekevin — CHANGES_REQUESTED** [review]

File: aptos-move/vm-genesis/src/lib.rs [hunk 20]
```diff
@@ -190,11 +192,9 @@ fn create_and_initialize_main_accounts(
 ) {
     let aptos_root_auth_key = AuthenticationKey::ed25519(aptos_root_key);
 
-    let genesis_gas_schedule = &INITIAL_COST_SCHEDULE;
-    let instr_gas_costs = bcs::to_bytes(&genesis_gas_schedule.instruction_table)
-        .expect("Failure serializing genesis instr gas costs");
-    let native_gas_costs = bcs::to_bytes(&genesis_gas_schedule.native_table)
-        .expect("Failure serializing genesis native gas costs");
+    let genesis_gas_params = AptosGasParameters::initial();
```
Comment (movekevin) [inline aptos-move/vm-genesis/src/lib.rs:195]: Are you going to make this configurable in the genesis blob in a followup PR?

File: aptos-move/vm-genesis/src/lib.rs [hunk 21]
```diff
@@ -360,7 +358,7 @@ fn publish_token_modules(
         let mut bytes = vec![];
         m.serialize(&mut bytes).unwrap();
         session
-            .publish_module(bytes, *module_id.address(), &mut GasStatus::new_unmetered())
+            .publish_module(bytes, *module_id.address(), &mut UnmeteredGasMeter)
```
Comment (movekevin) [inline aptos-move/vm-genesis/src/lib.rs:361]: What does this do?

File: aptos-move/vm-genesis/src/lib.rs [hunk 22]
```diff
@@ -562,7 +559,7 @@ pub fn test_genesis_module_publishing() {
     }
     let data_cache = StateViewCache::new(&state_view).into_move_resolver();
 
-    let move_vm = MoveVmExt::new().unwrap();
+    let move_vm = MoveVmExt::new(NativeGasParameters::zeros()).unwrap();
```
Comment (movekevin) [inline aptos-move/vm-genesis/src/lib.rs:562]: Does this make sense to charge no gas in tests though? Are you planning on adding e2e tests for gas costs in a followup PR?

File: types/src/on_chain_config/gas_schedule.rs [hunk 23]
```diff
@@ -0,0 +1,23 @@
...
+use serde::{Deserialize, Serialize};
+use std::collections::BTreeMap;
+
+#[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
+pub struct GasSchedule {
+    pub entries: Vec<(String, u64)>,
+}
+
+impl GasSchedule {
+    pub fn to_btree_map(self) -> BTreeMap<String, u64> {
```
Comment (movekevin) [inline types/src/on_chain_config/gas_schedule.rs:14]: How/when is this config reloaded from latest on chain state?

File: aptos-move/aptos-gas/Cargo.toml (hunk 10 above)
Comment (vgao1996) [reply aptos-move/aptos-gas/Cargo.toml:13]: `move-deps` force feature unification, which is causing some problems.

Comment (movekevin) [reply aptos-move/aptos-gas/Cargo.toml:13]: It does make it easier to upgrade Move though. Having more than one versions in the same repo seems a bit inconsistent to me and can potentially be dangerous if there are incompatibilities?

File: aptos-move/aptos-gas/src/gas_meter.rs (hunk 11 above)
Comment (vgao1996) [reply aptos-move/aptos-gas/src/gas_meter.rs:38]: Not totally sure what you mean. These are totally different gas parameters from different packages (move-stdlib, aptos-framework). The move-stdlib one in particular, is defined in the Move repo

File: aptos-move/aptos-gas/src/aptos_framework.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (vgao1996) [reply aptos-move/aptos-gas/src/aptos_framework.rs:6]: No this is not generated and should not be. The reason is that when we change some of the field names, we don't want the key to change or otherwise we break the on-chain gas schedule.

The macro also defines a few tests to ensure
1. All the paths (chain of field accesses) are unique
2. All keys are unique
3. All entries are covered

File: aptos-move/aptos-gas/Cargo.toml (hunk 10 above)
Comment (vgao1996) [reply aptos-move/aptos-gas/Cargo.toml:13]: > Having more than one versions in the same repo also seems a bit inconsistent to me and can potentially be dangerous if there are incompatibilities?

I think this is debatable. There are legitimate use cases to have different features enabled for different parts of the project. For example, developer tools like CLIs will want to include some optional testing features while the prod VM likely won't, due to security reasons.

Now you can argue that things like these don't necessarily need to be gated behind feature flags and I would agree with you, but unfortunately this is not how things work in the Move repo right now, and it would take us some time to talk to the community and make changes.

File: aptos-move/aptos-gas/src/instr.rs (hunk 12 above)
Comment (vgao1996) [reply aptos-move/aptos-gas/src/instr.rs:266]: Probably not worth it. The best you can do right now is to have
```
let (base_cost, unit_cost) = match op {
    COPY_LOC => (self.copy_loc_base, self.copy_loc_unit),
    ...
};
let cost = base_cost + unit_cost * unit_cost
```
which is as complicated as what I already have.

Plus, we'll want to accommodate for formulae that don't fit into this model. Due to some legacy reasons, `VEC_PUSH_BACK` is treated as a sized instruction, but its cost model should only consist a base cost. We'll likely want to refine some the formulae a bit more in the future, meaning that more gas parameters may be added.

File: aptos-move/aptos-vm/src/aptos_vm_impl.rs (hunk 13 above)
Comment (vgao1996) [reply aptos-move/aptos-vm/src/aptos_vm_impl.rs:53]: Yes

File: aptos-move/framework/aptos-framework/sources/configs/gas_schedule.move (hunk 15 above)
Comment (vgao1996) [reply aptos-move/framework/aptos-framework/sources/configs/gas_schedule.move:27]: > we might start setting timestamp early in the genesis flow, which will break all of these asserts

Well if that's the where we'll eventually want to be, then I'm fine removing this assertion.

Comment (vgao1996) [reply aptos-move/framework/aptos-framework/sources/configs/gas_schedule.move:27]: But maybe let's keep them for now so they are consistent with some other stuff like ChainId? You can remove them when you when you revisit the genesis flow later.

File: aptos-move/framework/aptos-framework/sources/configs/gas_schedule.move (hunk 18 above)
Comment (vgao1996) [reply aptos-move/framework/aptos-framework/sources/configs/gas_schedule.move:43]: See my comment above about assert_genesis().

File: aptos-move/framework/aptos-framework/sources/configs/gas_schedule.move (hunk 19 above)
Comment (vgao1996) [reply aptos-move/framework/aptos-framework/sources/configs/gas_schedule.move:53]: It used to be that any on-chain config change would require a reconfiguration. Not totally sure if this is still the case. @zekun000 any ideas?

File: aptos-move/vm-genesis/src/lib.rs (hunk 20 above)
Comment (vgao1996) [reply aptos-move/vm-genesis/src/lib.rs:195]: I assume it's already configurable in the aptos-gas crate? The numbers you see are for genesis.

File: aptos-move/vm-genesis/src/lib.rs (hunk 21 above)
Comment (vgao1996) [reply aptos-move/vm-genesis/src/lib.rs:361]: It's an dummy gas meter that does not meter anything. All charge operations will succeed.

File: aptos-move/vm-genesis/src/lib.rs (hunk 22 above)
Comment (vgao1996) [reply aptos-move/vm-genesis/src/lib.rs:562]: The vm-genesis crate is strictly a tool to help us generate a genesis blob. It does use the Move VM, but gas metering is 100% meaningless in this context since you want all operations to succeed.

Coming to the test itself. I think it's actually questionable whether this test makes sense or not.

File: types/src/on_chain_config/gas_schedule.rs (hunk 23 above)
Comment (vgao1996) [reply types/src/on_chain_config/gas_schedule.rs:14]: When you create an AptosVM, it will try to load the config from the state view

**Comment by vgao1996** [comment]
@movekevin I made some of the changes you requested to gas_schedule.move. Let me know if those changes look good to you and if I have answered all of your questions!

Again, thanks for the detailed review!

File: aptos-move/framework/aptos-framework/sources/configs/gas_schedule.move (hunk 17 above)
Comment (vgao1996) [reply aptos-move/framework/aptos-framework/sources/configs/gas_schedule.move]: Actually it looks we'll need this entry point at least before the new governance stuff come out. A smoke test currently uses this to set the gas schedule.

File: aptos-move/aptos-gas/src/gas_meter.rs (hunk 11 above)
Comment (vgao1996) [reply aptos-move/aptos-gas/src/gas_meter.rs:38]: Oh, maybe you are confused because the two expressions look exactly the same? Notice that `FromOnChainGasSchedule` is a trait, and the two fields have different types that each implements this trait.

File: aptos-move/aptos-gas/Cargo.toml (hunk 10 above)
Comment (davidiw) [reply aptos-move/aptos-gas/Cargo.toml:13]: That makes sense. Could you share what features you're trying to prevent from leaking in here? It seems like this might just be a move fast hack that we could iterate on later so that we translate aptos feature flags into move feature flags?

File: aptos-move/aptos-gas/src/gas_meter.rs (hunk 11 above)
Comment (davidiw) [reply aptos-move/aptos-gas/src/gas_meter.rs:38]: There's some Rust conversion magic as this calls a specific implementation based upon the destination type.

File: aptos-move/aptos-gas/src/transaction.rs [hunk 24]
```diff
@@ -0,0 +1,105 @@
...
+    [large_transaction_cutoff, "large_transaction_cutoff", 600],
+    // The units of gas that to be charged per byte over the `large_transaction_cutoff` in addition to
+    // `min_transaction_gas_units` for transactions whose size exceeds `large_transaction_cutoff`.
+    [intrinsic_gas_per_byte, "intrinsic_gas_per_byte", 8],
+    // ~5 microseconds should equal one unit of computational gas. We bound the maximum
+    // computational time of any given transaction at roughly 20 seconds. We want this number and
+    // `MAX_PRICE_PER_GAS_UNIT` to always satisfy the inequality that
+    // MAXIMUM_NUMBER_OF_GAS_UNITS * MAX_PRICE_PER_GAS_UNIT < min(u64::MAX, GasUnits<GasCarrier>::MAX)
+    // NB: The bound is set quite high since custom scripts aren't allowed except from predefined
+    // and vetted senders.
```
Comment (davidiw) [inline aptos-move/aptos-gas/src/transaction.rs:70]: This isn't true.

File: aptos-move/framework/aptos-framework/sources/configs/gas_schedule.move (hunk 19 above)
Comment (davidiw) [reply aptos-move/framework/aptos-framework/sources/configs/gas_schedule.move:51]: I think this is governance responsibilities to call.

File: aptos-move/aptos-gas/src/gas_meter.rs (hunk 11 above)
Comment (movekevin) [reply aptos-move/aptos-gas/src/gas_meter.rs:38]: Makes sense this also got me without looking at it in an IDE.

File: aptos-move/framework/aptos-framework/sources/configs/gas_schedule.move (hunk 15 above)
Comment (movekevin) [reply aptos-move/framework/aptos-framework/sources/configs/gas_schedule.move:27]: Sure

File: aptos-move/framework/aptos-framework/sources/configs/gas_schedule.move (hunk 17 above)
Comment (movekevin) [reply aptos-move/framework/aptos-framework/sources/configs/gas_schedule.move:40]: Makes sense

File: aptos-move/framework/aptos-framework/sources/configs/gas_schedule.move (hunk 19 above)
Comment (movekevin) [reply aptos-move/framework/aptos-framework/sources/configs/gas_schedule.move:51]: This can be done as a follow up but the part that's missing for me is how the Rust side syncs when the gas schedule changes on chain. If it's through reconfiguration, I think we should separate it out. Triggering reconfiguration can impact validator set and rewards calculation. We should avoid doing that for every single change if possible.

File: execution/executor-benchmark/src/transaction_generator.rs [hunk 25]
```diff
@@ -245,6 +246,8 @@ impl TransactionGenerator {
         block_size: usize,
         seed_account_balance: u64,
     ) -> Vec<Vec<Transaction>> {
+        println!("{}", seed_account_balance);
```
Comment (movekevin) [inline execution/executor-benchmark/src/transaction_generator.rs:249]: Leftover println!?

File: execution/executor-benchmark/src/transaction_generator.rs [hunk 26]
```diff
@@ -304,6 +307,8 @@ impl TransactionGenerator {
         init_account_balance: u64,
         block_size: usize,
     ) -> Vec<Vec<Transaction>> {
+        println!("{}", init_account_balance);
```
Comment (movekevin) [inline execution/executor-benchmark/src/transaction_generator.rs:310]: Same here

**Review by movekevin — APPROVED** [review]

File: types/src/on_chain_config/gas_schedule.rs (hunk 23 above)
Comment (movekevin) [reply types/src/on_chain_config/gas_schedule.rs:14]: Then it sounds like it shouldn't need reconfiguration to see gas schedule updates? Anyway this can be figured out in a separate followup PR as part of making sure gas schedule is upgradable :)

File: aptos-move/vm-genesis/src/lib.rs (hunk 20 above)
Comment (movekevin) [reply aptos-move/vm-genesis/src/lib.rs:195]: Hmm that's a fair point. I was thinking of having different gas schedules for different testnets, but probably not a use case in demand yet as we want testnet to be as realistic as possible anyway so it makes sense to keep them the same for now.

File: aptos-move/framework/aptos-framework/sources/configs/gas_schedule.move (hunk 19 above)
Comment (zekun000) [reply aptos-move/framework/aptos-framework/sources/configs/gas_schedule.move:51]: it's the correct use of the new epoch event/reconfiguration because it needs to trigger a new epoch (for validators to align on the same gas schedule)

File: aptos-move/vm-genesis/src/lib.rs (hunk 20 above)
Comment (vgao1996) [reply aptos-move/vm-genesis/src/lib.rs:195]: In addition to what you mentioned, I also think that tests should have their own gas schedule defined. Otherwise they can be quite sensitive to gas schedule changes and it would be extremely hard to debug.

File: aptos-move/aptos-gas/src/transaction.rs (hunk 24 above)
Comment (vgao1996) [reply aptos-move/aptos-gas/src/transaction.rs:70]: It's an old comment I copied over. Will get it deleted

File: aptos-move/aptos-gas/Cargo.toml (hunk 10 above)
Comment (vgao1996) [reply aptos-move/aptos-gas/Cargo.toml:13]: > Could you share what features you're trying to prevent from leaking in here?
If move-cli is present in the build tree, then move-stdlib's "testing" feature will be enabled, even if you don't specify it.

> It seems like this might just be a move fast hack that we could iterate on later so that we translate aptos feature flags into move feature flags?
The problem is in the Move repo and I need to think about it more.
